File: client/src/types/station.ts
export type RiskCategory = 'Safe' | 'Semi-Critical' | 'Critical' | 'Over-Exploited';

export interface Station {
  id: string;
  station_code: string;
  name: string;
  latitude: number;
  longitude: number;
  district: string;
  state: string;
  block?: string;
  aquifer_type?: string;
  well_depth_m?: number;
  current_risk_category?: RiskCategory;
  current_depth_m?: number;
  months_to_next_risk_tier?: number;
  forecast_risk_category?: RiskCategory;
}

export interface Reading {
  timestamp: string;
  depth_below_ground_m: number;
  quality_flag: string;
}

export interface StationWithReadings extends Station {
  readings: Reading[];
  created_at: string;
  updated_at: string;
}

export interface ForecastPoint {
  date: string;
  projected_depth_m: number;
}

export interface StationForecast {
  station_id: string;
  station_code: string;
  current_risk_category?: RiskCategory;
  current_depth_m?: number;
  forecast_risk_category?: RiskCategory;
  months_to_next_risk_tier?: number;
  trend_direction: 'declining' | 'stable' | 'recovering';
  rate_of_change_m_per_year: number;
  forecast_points: ForecastPoint[];
  confidence: 'low' | 'moderate' | 'high';
  data_points_used: number;
}

export interface DistrictSummary {
  district: string;
  state: string;
  total_stations: number;
  safe_count: number;
  semi_critical_count: number;
  critical_count: number;
  over_exploited_count: number;
  unclassified_count: number;
}

export interface Advisory {
  station_id: string;
  station_code: string;
  risk_category?: RiskCategory;
  summary: string;
  recommendation: string;
  urgency: 'low' | 'moderate' | 'high' | 'critical';
  generated_at: string;
}

export const RISK_COLORS: Record<RiskCategory, string> = {
  'Safe': '#22c55e',
  'Semi-Critical': '#f59e0b',
  'Critical': '#f97316',
  'Over-Exploited': '#ef4444',
};

export const RISK_ORDER: RiskCategory[] = [
  'Safe',
  'Semi-Critical',
  'Critical',
  'Over-Exploited',
];

File: server/app/schemas/station.py
"""Pydantic schemas for Station API responses."""

from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field

from app.core.constants import RiskCategory


# ── Base ──────────────────────────────────────────────────────────────────────

class StationBase(BaseModel):
    station_code: str
    name: str
    latitude: float
    longitude: float
    district: str
    state: str
    block: Optional[str] = None
    aquifer_type: Optional[str] = None
    well_depth_m: Optional[float] = None


class StationSummary(BaseModel):
    """Lightweight station representation for list views and map markers."""
    id: UUID
    station_code: str
    name: str
    latitude: float
    longitude: float
    district: str
    state: str
    current_risk_category: Optional[RiskCategory] = None
    current_depth_m: Optional[float] = None
    months_to_next_risk_tier: Optional[int] = None

    model_config = {"from_attributes": True}


class StationDetail(StationBase):
    """Full station detail including forecast info."""
    id: UUID
    current_risk_category: Optional[RiskCategory] = None
    current_depth_m: Optional[float] = None
    months_to_next_risk_tier: Optional[int] = None
    forecast_risk_category: Optional[RiskCategory] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Readings ──────────────────────────────────────────────────────────────────

class ReadingOut(BaseModel):
    timestamp: datetime
    depth_below_ground_m: float
    quality_flag: str = "valid"

    model_config = {"from_attributes": True}


class StationWithReadings(StationDetail):
    """Station detail with its full time-series of readings."""
    readings: List[ReadingOut] = []


# ── Forecast ──────────────────────────────────────────────────────────────────

class ForecastPoint(BaseModel):
    """A single point in the projected trend line."""
    date: datetime
    projected_depth_m: float


class StationForecast(BaseModel):
    """Forecast response for a single station."""
    station_id: UUID
    station_code: str
    current_risk_category: Optional[RiskCategory] = None
    current_depth_m: Optional[float] = None
    forecast_risk_category: Optional[RiskCategory] = None
    months_to_next_risk_tier: Optional[int] = None
    trend_direction: str = Field(description="'declining' | 'stable' | 'recovering'")
    rate_of_change_m_per_year: float = Field(description="Meters per year depth change")
    forecast_points: List[ForecastPoint] = []
    confidence: str = Field(default="moderate", description="'low' | 'moderate' | 'high'")
    data_points_used: int = 0


# ── District ──────────────────────────────────────────────────────────────────

class DistrictSummary(BaseModel):
    """Aggregate risk summary for a district."""
    district: str
    state: str
    total_stations: int
    safe_count: int = 0
    semi_critical_count: int = 0
    critical_count: int = 0
    over_exploited_count: int = 0
    unclassified_count: int = 0


# ── Advisory ──────────────────────────────────────────────────────────────────

class AdvisoryResponse(BaseModel):
    """LLM-generated advisory for a station."""
    station_id: UUID
    station_code: str
    risk_category: Optional[RiskCategory] = None
    summary: str = Field(description="One-paragraph plain-language situation summary")
    recommendation: str = Field(description="Actionable recommendation for stakeholders")
    urgency: str = Field(default="moderate", description="'low' | 'moderate' | 'high' | 'critical'")
    generated_at: datetime


# -- Pagination ----------------------------------------------------------------

class PaginatedStations(BaseModel):
    """Paginated list of station summaries."""
    items: List[StationSummary] = []
    total: int = Field(description="Total number of matching stations")
    page: int = Field(description="Current page number (1-indexed)")
    page_size: int = Field(description="Number of items per page")
    total_pages: int = Field(description="Total number of pages")
